src/scenarios/shared/runner.py

```diff
@@ -573,11 +575,87 @@ def run(self):
+                    original_traced_enable = None
+
+                    try:
+                        # Get the current value of persist.traced.enable
+                        getLogger().info("Getting current persist.traced.enable value")
+                        get_traced_cmd = xharness_adb() + [
+                            'shell',
+                            'getprop persist.traced.enable'
+                        ]
+                        get_traced_result = RunCommand(get_traced_cmd, verbose=True)
+                        get_traced_result.run()
+                        original_traced_enable = get_traced_result.stdout.strip()
+
+                        # Setup the phone props to allow perfetto to run properly
+                        getLogger().info("Setting up the device for Perfetto")
+                        setup_perfetto_cmd = xharness_adb() + [
+                            'shell',
+                            'setprop persist.traced.enable 1'
+                        ]
+                        RunCommand(setup_perfetto_cmd, verbose=True).run()
+
+                        getLogger().info("Tracing with Perfetto")
+                        # Get the max TotalTime from the allResults list in seconds
+                        max_startup_time_sec = int(max(int(re.search(r"TotalTime: (\d+)", str(result)).group(1)) for result in allResults) / 1000)
+                        perfetto_max_trace_time_sec = max_startup_time_sec * 2 # Set the max trace time to be double the max startup time
+                        if max_startup_time_sec > 60:
+                            getLogger().error(f"Max startup time is greater than 60 seconds (Max startup time: {max_startup_time_sec}), this means something probably went wrong.")
+                            raise Exception("Max startup time is greater than 60 seconds, this means something probably went wrong.")
+
+                        perfetto_cmd = xharness_adb() + [
+                            'shell',
+                            f'perfetto --background --txt -o {perfetto_device_save_file} --time {perfetto_max_trace_time_sec}s -b 64mb sched freq idle am wm gfx view binder_driver hal dalvik camera input res memory'
+                        ]
+                        RunCommand(perfetto_cmd, verbose=True).run()
+
+                        # Run the startup test with the trace running (only once)
+                        getLogger().info("Running startup test with Perfetto trace running")
+                        traced_start = RunCommand(androidHelper.startappcommand, verbose=True)
+                        traced_start.run()
+
+                        # Stop perfetto now that the app has started. Sending a Terminate signal should be enough per the longer trace capturing guidance here: https://perfetto.dev/docs/concepts/config#android.
+                        getLogger().info("Stopping perfetto trace capture")
+                        stop_perfetto_cmd = xharness_adb() + [
```

Comment:
Should the perfetto killall also be in the finally block?

Reply:
Additionally, it might make sense to run a precautionary perfetto killall before running the kill command at the top. I know a couple times we have been bitten by processes not being killed.

Reply:
Having a kill before the execution is a great idea. I have updated the code per these comments. I included a perfetto kill in the finally block, though am not super worried about it as the time parameter in the call starting the trace should ensure it ends eventually.